src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -922,6 +922,8 @@ void MacroAssembler::li(Register Rd, int64_t imm) {
   }                                                                                   \
   void MacroAssembler::NAME(Register Rd, Label &L, Register temp) {                   \
     assert_different_registers(Rd, temp);                                             \
+    /* We can't patch C, i.e. if Label wasn't bound we need to patch this jump.*/     \
+    IncompressibleRegion ir(this);                                                    \
```

Comment:
Seems that we have rare `jal` with unbound label. Here is what I see it. 
For unbound label, we set `dest` to pc() at [1]. So the `distance` calculated at [2] will always be zero for this case. 
After that we emit a simple `Assembler::jal(Rd, 0);` at [3] which is there for patching. So no `jalr` involved. 
I think this means that possible future callers of `jal` with unbound label have to ensure that the target is not far. 

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L947
[2] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L913
[3] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L915

Reply:
Then I suggest we remove jalr from the method.

Reply:
Note, I notice this when I enable c_jal, if we ever want c_jal we need to either eanble patching of C or turn off C when we need patching.

Reply:
> Then I suggest we remove jalr from the method.

I think we still need the `jalr` which is emitted by the else branch of the method. The reason is that we have calls to `void jal(const address dest, Register temp = t0);` where we need to check whether the dest is far or not [1][2].

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp#L1815
[2] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L4371

Reply:
> Note, I notice this when I enable c_jal, if we ever want c_jal we need to either eanble patching of C or turn off C when we need patching.

The current approach is to turn off C where patching could happen. There are two relocate assembler functions for that purpose [1]. They will turn off C on those patching sites.

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/assembler_riscv.hpp#L2083-L2093

Reply:
Those two plain jal don't use Labels, so they are unaffected.

In this case, instead of turn off C we have no C switch for jal.
So we don't have one approach. Turn off C would be inline with what you are saying.

Reply:
Sorry I was talking about C.J. As that is the one which maps to JAL uncond jumps. (x0)
It's not used at all as it is now.